linux/src/threadpool_linux.c

```diff
@@ -501,29 +531,156 @@ int threadpool_schedule_work(THREADPOOL_HANDLE threadpool, THREADPOOL_WORK_FUNCT
-    (void)work_function_ctx;
-    (void)timer_handle;
-    return MU_FAILURE;
+    int result;
+
+    if (
+        threadpool == NULL ||
+        work_function == NULL ||
+        timer_handle == NULL
+        )
+    {
+        LogError("Invalid args: THREADPOOL_HANDLE threadpool = %p, uint32_t start_delay_ms = %" PRIu32 ", uint32_t timer_period_ms = %" PRIu32 ", THREADPOOL_WORK_FUNCTION work_function = %p, void* work_function_context = %p, TIMER_INSTANCE_HANDLE* timer_handle = %p",
+            threadpool, start_delay_ms, timer_period_ms, work_function, work_function_ctx, timer_handle);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        (void)interlocked_increment(&threadpool->pending_call_count);
+
+        TIMER_INSTANCE* timer_instance = malloc(sizeof(TIMER_INSTANCE));
+        if (timer_instance == NULL)
+        {
+            LogError("Failure allocating Timer Instance");
+        }
+        else
+        {
+            timer_instance->work_function = work_function;
+            timer_instance->work_function_ctx = work_function_ctx;
+
+            // Setup the timer
+            struct sigevent sigev = {0};
+            timer_t time_id = 0;
+
+            sigev.sigev_notify          = SIGEV_THREAD;
+            sigev.sigev_notify_function = on_timer_callback;
+            sigev.sigev_value.sival_ptr = timer_instance;
+
+            if (timer_create(CLOCK_REALTIME, &sigev, &time_id) != 0)
+            {
+                char err_msg[128] = {0};
+                (void)strerror_r(errno, err_msg, 128);
+                LogError("Failure calling timer_create. Error: %s", MU_P_OR_NULL(err_msg));
+            }
+            else
+            {
+                struct itimerspec its = {0};
+                its.it_value.tv_sec = start_delay_ms / 1000;
+                its.it_value.tv_nsec = start_delay_ms * MILLISEC_TO_NANOSEC % 1000000000;
+                its.it_interval.tv_sec = 0;
+                its.it_interval.tv_nsec = timer_period_ms * MILLISEC_TO_NANOSEC;
+
+                if (timer_settime(time_id, 0, &its, NULL) == -1)
+                {
+                    char err_msg[128] = {0};
+                    (void)strerror_r(errno, err_msg, 128);
+                    LogError("Failure calling timer_settime. Error: %s", MU_P_OR_NULL(err_msg));
+                }
+                else
+                {
+                    (void)interlocked_decrement(&threadpool->pending_call_count);
+                    wake_by_address_all(&threadpool->pending_call_count);
+                    timer_instance->time_id = time_id;
+                    *timer_handle = timer_instance;
+                    result = 0;
+                    goto all_ok;
+                }
+            }
+            free(timer_instance);
+        }
+        result = MU_FAILURE;
+        (void)interlocked_decrement(&threadpool->pending_call_count);
+        wake_by_address_all(&threadpool->pending_call_count);
+    }
+all_ok:
+    return result;
```

Comment:
>return result; [](http://example.com/codeflow?start=4&length=14)

I've read all of this function. I fail to see ANY connection between `threadpool` and the timer. Effectively starting 100 of the timers, will create 100 of the `timer_t`. 

`timer_t` will fire on "some thread", not on one of the threads of the threadpool.

Note: `SIGEV_THREAD
              Upon timer expiration, invoke sigev_notify_function as if it were the start function of a  new  thread.
              See sigevent(7) for details.`

I would expect SOME limitation of the thread executions here.

As in... if the threadpool was created with some `max_thread_count`... then... we'd better not go over that limit.

Reply:
In Win I believe there is a connection between the timer and the threadpool.  In linux they are separate entities.  The system takes care of the timer and the threadpool is controlled by us.  So this is just another instance of this API being written for windows.